src/tools/clippy/tests/ui/needless_lifetimes.rs

```diff
@@ -527,7 +527,7 @@ mod in_macro {
 }
 
 mod issue5787 {
-    use std::sync::MutexGuard;
+    struct MutexGuard<'a, T>(std::marker::PhantomData<fn() -> &'a T>);
```

Comment:
What's the motivation of this change?

Reply:
Thanks for the reminder! rust-lang/rust-clippy#5787 is about unnecessary lint on the `'a` lifetime in the `Foo::wait` signature below. By putting `MutexGuard` onto the coroutine variant instead, clippy now lints loudly that `MutexGuard` is held across `await`. This is technically not related to issue 5787.

However, I think I should add a test in `src/tools/clippy/tests/ui/await_holding_lock.rs` for this case, because holding a `MutexGuard` from `std` in an upvar is equally bad and shall be linted. Even if the guard is dropped at the beginning of the async body, it will be a problem if the future is instantiated but not polled. I am working on this clippy test.

Reply:
I see. In this case we usually just allow the second lint that triggers unexpectedly. I would prefer if you add an `allow(await_holding_lock)` here, over re-defining `MutexGuard`. That would be way easier to understand and explain to a reader of this test case.

Reply:
The new test is in.